### ISSUE TITLE
The default priority of a pushed request

### ISSUE THREAD

**Issue by kazuho** [httpwg/http-extensions#1056]
The answer depends on if reprioritization should be part of the spec (i.e. resolution of #1021), and if browsers would use it.

For more context, see [this thread](https://github.com/httpwg/http-extensions/pull/1048#discussion_r376009024).

**Comment by LPardue** [comment]
There was some previous discussion at https://github.com/kazuho/draft-kazuho-httpbis-priority/issues/91, lets continue the conversation here.

**Comment by MikeBishop** [comment]
If the pushed resource gets assigned a priority that means it won't get sent until the client sends the desired priority, that's effectively no different than a Link header triggering a request for the resource.  So assigning it the minimum priority is essentially useless unless the connection is otherwise idle.  But if given a priority where it *can* be sent, the resource likely can be pushed and already sent within the 1+ RTT it will take for the client to see the promise and reprioritize it.  To me, that indicates we can't rely on reprioritization of pushes, regardless of whether that feature remains in the spec.

The HTTP/2 approach of "slightly less important than the document it was pushed on" seems like a more sensible starting point.  In this incarnation, that probably means the same priority level, but not incremental, regardless of what the parent was.  Or perhaps one level lower, to ensure that the parent can always pre-empt the push if it becomes available in mid-transfer.

**Comment by tombergan** [comment]
For any given push, the correct priority is dependent on context. There is no perfect answer that will work optimally in all cases. In deciding a default, we can either (a) choose something we think is closest to correct for most cases, or (b) choose something that will minimize harm in most cases. I favor (b). It's very difficult to get (a) right for all of the reasons discussed in #1026.

Push is often speculative: the server predicts that the client may need a resource, but the server doesn't know for sure (the resource might already be cached, or maybe it's not needed at all). "Over pushing" is a known problem and we have observed that push can be a footgun. Push is actually very hard to use well. @MikeBishop is concerned that assigning the minimum priority makes push useless unless the connection is otherwise idle, but that's exactly the point! If a push is speculative, as they often are in my experience, then idle times are the safest times to push.

Further, I would argue that no matter which priorities are chosen, push is very rarely useful on connections that are never idle. Suppose a web page has three files, HTML, CSS, and JPG, prioritized in that order. If the HTML is large enough (and can be generated quickly enough) so that the connection is never idle, then it's never beneficial to push the CSS or the JPG. The HTML is hogging the connection, so you might as well wait for the client to request each file explicitly. I can think of just two cases where push might be useful on a connection that is never idle:

1. The server wants to interleave a pushed CSS file immediately after sending `<link rel=stylesheet>` from the HTML. The server will need custom logic for this anyway. The default priority is irrelevant.

2. The client asks for a web page concurrently with another stream of requests. I don't have a good solution for this. It's difficult for the server to know what the correct priorities should be for a pushed resource relative to other concurrent requests (see #1026). I also believe this is an uncommon case -- the most popular motivating example for push involves a user who has never been to your site previously.

@MikeBishop, I didn't grok what you meant by "we can't rely on reprioritization of pushes"?

**Comment by MikeBishop** [comment]
I believe client-driven reprioritization of pushes will be useless because it takes 1 RTT for the client to receive the promise and respond with a reprioritization directive.  There are only a few realistic cases:
- The default priority is low enough that the resource has not yet begun to be sent by the time reprioritization arrives.  Thus, the push was pointless and we should have just waited for the client to make a request.
- The default priority is high enough that the resource has already been encrypted and sent by the time reprioritization arrives.  Thus, the reprioritization was useless because the data has already been committed.
- The default priority is high enough that the resource has already been encrypted and sent by the time reprioritization arrives, *and* the resource is large enough that it's still being streamed when the client's reprioritization arrives.  We already know that pushing large resources tends to hurt more than it helps, so servers shouldn't do this anyway.

It's a fair point that push should primarily utilize idle connection time, and from that standpoint, it doesn't matter what priority it's assigned.  It might hypothetically be competing with things the client has prioritized as only being sent when the connection is idle, but that should be uncommon.

**Comment by LPardue** [comment]
For some population of intermediaries, the action of pushing is driven using a Link "rel=preload" in the response of the first request. At that time, the intermediary might have very little information about the resource when it synthesizes the push; there may be an `as` parameter but the size is unlikely to be known and so the server can't execute some of the smarts suggested in case 3.

The interplay with push and request generation order is interesting, in section 3.2 of the editor's draft we currently say

```
Serving non-incremental responses one by one, in the order in which those requests were generated is considered to be the best strategy.
```

If we are pushing at "one below the parent request priority" non-incrementally, then there is a danger that these block real client requests at or below the same level. Pushing at too high a priority level seems to endanger client's actual authority over their usage of the priority levels - even if everyone is acting in good faith.

**Comment by tombergan** [comment]
> It's a fair point that push should primarily utilize idle connection time, and from that standpoint, it doesn't matter what priority it's assigned.

The default priority doesn't matter in case 2, I agree with that, but it matters in cases 1 and 3.

For cases 1 and 3, it's not enough to simply state that push should not have been used. While I mostly agree with that in theory, in practice it's very difficult to know which case you'll end up in a priori. Suppose the user requests A and the server considers pushing B. The server cannot predict which case it will end up in unless it precisely knows (a) how long it will take to generate the responses for A and B (to know if there will be idle time before sending the first byte of A); (b) the sizes of A and B (see Lucas' response); and (c) the effective BDP of the connection (possible to estimate in theory but this assumes the server push code has access to the current cwnd).

In practice, even well-meaning servers are going to use push in cases 1 and 3. This implies that the default priority will be important in practice, and even "one priority below the parent request" has problems as Lucas points out.

**Comment by kazuho** [comment]
Revisiting this issue, I am starting to wonder if we need to provide an advice. As others have noted, how a pushed resource should be prioritized depends on each use-case. And for the case of an H3 terminator not having the knowledge of how to prioritize a pushed response, the priority response header from the upstream server can provide an answer.

To summarize, maybe saying something like: _When pushing a response, it is the responsibility of the server to correctly prioritize that response against other responses. When making that decision, an HTTP server can take into consideration the value of the priority response header field of the pushed response._

**Comment by LPardue** [comment]
I've lost some context here. Looking back through the threads, in https://github.com/kazuho/draft-kazuho-httpbis-priority/issues/91#issuecomment-555416246  you (Kazuho) said:

> When push is used, both the request and response headers are created by the server. If either of the two contains a priority header field, the value of that header field should be used.
> The question is what to do when there is no hint.

This seems to conflict with your proposed advice.

**Comment by kazuho** [comment]
@LPardue Thank you for resurrecting my old comment.

I think I am consistent in arguing that a server _can_<sup>1</sup> use a hint provided by upstream, if such hint is provided.

What I am arguing now is that we should encourage servers determine the priority of the pushed request, rather than trying to find a default. As we have discussed, one default does not fit all.

When a server is the initiator of a pushed request, it would definitely know why it initiated that push. As it knows the reason, it would be possible to determine the correct urgency. When a server is obeying to an instruction from an upstream server (e.g., link rel=preload), the upstream server can provide a hint by using the priority header field in the pushed response, or make a guess based on other properties (e.g., type of the pushed resource, the fact that link rel=preload is being used, etc.).

These hints could work better than a hard-coded default in the spec. Assuming that is the case, my point is that we should encourage the use of such hints.

[1] Using "can" rather than "SHOULD", as we've reduced the guidance (see #1026).

**Comment by LPardue** [comment]
Encouraging signals is good and I agree that without them bets are off. I think there is ultimately no "good" guidance to provide, so maybe all we can do is highlight the pitfalls.

Writing this I realise a pitfall I'll need to confirm. Right now, the quiche implementation factors in the stream ID into scheduling choices. This might break down if the request and push stream IDs start to diverge or progress at different phases. This is a manifestation of my earlier comment about considering request order for scheduling.

**Comment by LPardue** [comment]
In review of the PR @rmarx raises the point that he believes there should be more concrete guidance. Especially wrt a default for pushes in the web browsing use case.

There seems to be a popular opinion that push for Web browsing is going the way of the Dodo. That doesn't make me super inclined to try an address one specific case, when this scheme is purposefully for general HTTP.

For me, defaults are about describing the "null extensible priority signal" and what that means as  input criteria to scheduling. Clients that omit Prioirity for requests in this scheme are making an explicit choice. Pushed requests are weird, with no standard defined internal API and underspecified interop (at least in HTTP land). Whomever generates a push on the server has 2,  well-coordinated, opportunities to provide a signal into the prioiritization process. There might be reason why those opportunitiesare not taken but I think we should discourage it by highlighting that there is no push default. Server has to make a decision, which is left as an implementation choice. The PR attempts to list some important factors to consider in doing so.

The only thing I might be tempted to add to the PR is something like

"A server might want to prioiritize a pushed response while ignoring all other important factors. Such a naive implementation can apply the lowest urgency in order to not compete for resources with other concurrent activity. However, this risks late delivery of the push, which might undermine its intended goal."

**Comment by MikeBishop** [comment]
Given that the server generates the request as well as the response, perhaps a recommendation that the PUSH_PROMISE headers contain the prioritization that the server intends to apply?  If the client dislikes the indicated priority, it can send an update.  That means there effectively _is_ a default, but what that default is remains in the hands of the server.

**Comment by LPardue** [comment]
That has some nice theoretical purity. I'm a bit skeptical that people would actual apply this recommendation, the response header could also contain that same information, so we're talking a thin margin of opportunity to reprioritize a push. Upthread we've discussed how servers may already have commited data to unrevokable buffers. Is there really interest adding complexity for this margin? For simplicity sake, would it suffice to recommend that servers send the response header in order to give the client a signal it can use to build some facsimile of a world view?

**Comment by MikeBishop** [comment]
By the point you're getting response headers, you're probably getting data next, unless we expect servers to generate 103s for pushes too.  At that point, the only "world view" that matters is what's actually coming down the pipe.  The promise might not have much lead time, but I suspect it's the best we get; if the server fires off a batch of promises and then starts fulfilling them sequentially, for example.

**Comment by LPardue** [comment]
Right. But to give an example of my problem as a CDN operator, I might consume a link rel=preload header to trigger a push promise. At that stage I actually have no idea what to priortize the response data at, because I have limited knowledge of what the response is. I'd prefer to wait for the origin's response, which could include a priority signal, that will definitely fill in my knowledge. Sending any value in PUSH_PROMISE is a guessing game IMO. I suspect in HTTP/3 it might also have a weird effectif a server tries to promise the same thing across multiple resources, but I have not thought that through much.

There is a window of opportunity to say "well I might do this with some degree of uncertainty" and a client to respond with "well if you do then here's a hint of what I want, which might conflict with the actual prioritization that gets applied". I find it hard to see how anyone would see there being enough confidence in this optimisation to really care to implement. Bit I'm happy to be wrong here.

**Comment by MikeBishop** [comment]
That's true, but I'd assert that as a CDN operator, if you don't already have it locally you shouldn't be trying to push it in the normal case.  Push works best to fill idle time, and if you have to fetch it, there won't be any.  Maybe use it to pre-warm your cache instead.

Maybe this works best as a statement of possibility -- "the server can include the prioritization it intends to apply as a request or response header."

**Comment by LPardue** [comment]
This obviously depends on the CDN architecture and the locality of the response to the component terminating the H2 or H3 connection :) I'll just say that if customers send a preload hint to the CDN edge and expect a push to occur, they might complain if the push didn't get promised, regardless whether it was a sub-optimal thing to do or not. 

This is just one unfortunate example that illustrates my concern about heaping more expectation and complexity to push.

Statement of possibility sounds fine though, let's see if I can we've that into the PR.

**Comment by kazuho** [comment]
While I agree that we can give more advice re how to push for web browsing use-case, I am not sure if that is going to help, when web browsers are deprecating their support for push.

In addition, I would point out that the intended audience to whom we think giving advice is beneficial (i.e. the servers not being that intelligent), are the ones that suffer most from misusing push. If we are going to provide any guidance to those, I think it is better to discourage the use of push entirely.

**Comment by rmarx** [comment]
I would agree with Kazuho here that the non-intelligent servers shouldn't employ push. This is however imo a weird thing to say in the priorities draft, which is why I suggested setting a low/lowest default push priority, effectively making sure it doesn't impact normal behaviour (or impacts it less) if abused in a browsing use case. If we do consider Push mostly usable in a non-browser setup going forward, not specifying a default is probably fine, as you'll have much different client-side logic as well, which requires different tuning either way. 

I'm also not sure about Mike's suggestion about having PUSH_PROMISE reflect the server's idea of urgency for that resource, a) for the reasons Lucas listed and b) because I don't particularly believe in smart re-prioritization of Push (early on normal prioritization for other resources should take over, later on (e.g., pushing subresources of a REST response) you'll probably have much less contention from client-side requested resources).

**Comment by kazuho** [comment]
@rmarx 
> I would agree with Kazuho here that the non-intelligent servers shouldn't employ push. This is however imo a weird thing to say in the priorities draft, which is why I suggested setting a low/lowest default push priority, effectively making sure it doesn't impact normal behaviour (or impacts it less) if abused in a browsing use case.

Maybe we could start the paragraph by pointing out the difficulty of using push correctly, then go like "but if you think you are clever enough, use whatever information you have to the correct thing."
